### ISSUE TITLE
Why do I get a 'You have a leaked pooled object' warning?

### ISSUE THREAD

**Issue by hherrada** [rstudio/pool#21]
I have a MySQL DB running on Ubuntu and am following [this](http://shiny.rstudio.com/articles/pool-basics.html) article. Any insight will be highly appreciated! Thanks!

**Comment by bborgesr** [comment]
The "leaked pooled object" warning is meant to appear when you check out an object from the pool (e.g. `poolCheckout(pool)`), lose any reference to it, but never return it... Once `pool` detects this has happened, it will destroy the object, but it will also warn to let you know about it (this may be good info to have if you're trying to do some performance optimization, for example). 

However, if you're just using direct pool methods (i.e. you never check out a connection directly and always use statements like `dbGetQuery(pool, ...)`), this should not happen. 

If that's the case, could you do the following for me:

1. Make sure you're using the latest `pool` version.
2. Give me a reproducible example if you can (feel free to use the DB I have in the article).

**Comment by hherrada** [comment]
Hi! Thanks for the reply. I was putting an example together and ran into this error. Do you know what's wrong? I did install both packages from the dev repositories (DBI Version: 0.5-12 and pool Version: 0.1.0).

Thank you!

`library(DBI)`
`library(pool)`

`pool <- dbPool(
  drv = RMySQL::MySQL(),
  dbname = "shinydemo",
  host = "shiny-demo.csa7qlmguqrf.us-east-1.rds.amazonaws.com",
  username = "guest",
  password = "guest"
)`

`dbGetQuery(pool, "SELECT * FROM City LIMIT 5;")`

`Error in value[[3L]](cond) : `
`Object could not be returned back to the pool. It was destroyed instead. Error message: unable to find an inherited method for function ‘dbListResults’ for signature ‘"MySQLConnection"’`

**Comment by XtremeCurling** [comment]
@hherrada I was getting nearly the same error as you when trying to use `poolReturn`:

`Error in value[[3L]](cond) :
Object could not be returned back to the pool. It was destroyed instead. Error message: unable to find an inherited method for function ‘dbIsValid’ for signature ‘"PostgreSQLResult"’`

After some digging, I discovered this: not all database system drivers are created equal. `RPostgreSQL` might implement some methods from `DBI` that `RMySQL` doesn't, and vice-versa.

For me, the problem was that there's no `dbIsValid` method in the `RPostgreSQL` package. `poolReturn` calls `onPassivate`, which in turn calls `dbIsValid`, which isn't a defined method for the `PostgreSQLResult` signature, hence the error when calling `poolReturn`. I discovered that this method isn't defined by searching the `RPostgreSQL` package documentation.

I checked `RMySQL`, though, and was puzzled to find that there _is_ a `dbListResults` method defined for the `MySQLConnection` signature. So, not sure where your problem is coming from. Just thought I'd elucidate things a bit for you or anyone else running into this type of issue.

**Comment by bborgesr** [comment]
Thank you both @hherrada and @XtremeCurling for catching and sharing this. I'm sorry I haven't been super responsive lately. It does seem like some packages (DBI, RMySQL) have been changing in a such a way that they no longer play nice with pool devel (see #25). I hope to sort this all out properly soon, but for now, a way around this is to use the CRAN versions of both DBI and RMySQL when you want to use them with pool.

**Comment by robsalasco** [comment]
any news on this?

**Comment by bborgesr** [comment]
@robsalasco this shouldn't be happening anymore.... Can you give me a reproducible example of your current problem? The reason I haven't closed this issue yet is because DBI devel introduced breaking changes from DBI on CRAN, and I'm waiting for all of that to stabilize before looking into these issues again. However, I did remove the calls to `dbListResults`, so that you shouldn't see this warning message `'You have a leaked pooled object'` whenever you're using direct pool methods.

**Comment by robsalasco** [comment]
```
library(tidyverse)
library(RMySQL)
library(pool)

pool <- dbPool(drv = RMySQL::MySQL(),
dbname = "shinydemo",
host = "shiny-demo.csa7qlmguqrf.us-east-1.rds.amazonaws.com",
username = "guest",
password = "guest")

test1 <- src_pool(pool) %>% 
  tbl("City") %>% 
  select(Name,Population) %>% 
  collect(n=Inf)

test2 <- src_pool(pool) %>% 
  tbl("Country") %>% 
  select(Name,Continent,Population) %>% 
  collect(n=Inf)
```

I'm using the lastest development versions of dplyr and pool!

**Comment by bborgesr** [comment]
Right, I should have made that clear. The posts above were only using `DBI` directly (and not `dplyr` like you). The warning does not appear when you just use `DBI`. However, you are right, the integration with `dplyr` is still not perfect. That should also be fixed in the near future (within the next couple of months). The `dplyr` package already has merged a PR that allows this fix; now, it's a matter of that fix hitting CRAN (during the next release of `dplyr`) and us doing the necessary work in `pool` (started in this [branch](https://github.com/rstudio/pool/tree/dplyr-master), but not yet finished). So if you can afford a few months for the fix, this warning should disappear without you having to do anything else. If you can't afford it, you can either use `pool` + `DBI` or then `dplyr` directly (no `pool`), depending on your needs. I'm sorry this is such a pain... Since `pool` lies at the intersection of many other packages, it's hard to keep it all synced up...

**Comment by robsalasco** [comment]
Hi Barbara! I found a workaround for this 

```
library(tidyverse)
library(RMySQL)
library(pool)

pool <- dbPool(drv = RMySQL::MySQL(),
dbname = "shinydemo",
host = "shiny-demo.csa7qlmguqrf.us-east-1.rds.amazonaws.com",
username = "guest",
password = "guest")

con <- src_pool(pool)

test1 <- tbl(con,"City") %>% 
  select(Name,Population) %>% 
  collect(n=Inf)

test2 <- tbl(con,"Country") %>% 
  select(Name,Continent,Population) %>% 
  collect(n=Inf)

```

The warning now is gone!

**Comment by bborgesr** [comment]
Right, that's expected behavior. What you're doing in that case is actually directly checking out a connection from the pool and using that throughout (instead of using `pool` directly for the queries, and letting the pool decide which connection to use). That can be a fine approach if you're reasonably certain the connection holds. However, if you're doing that, you might as well not use `pool` (and just use `dplyr` directly), because you're not reaping any of the benefits of `pool`. But if understand the difference between the two approaches and you're okay with this one, there's absolutely no harm is doing this way (it just doesn't bring you any benefit either).

**Comment by bborgesr** [comment]
All issues on this thread should be solved now (the dplyr-related issues should be closed as of #42)
